crypto/pkcs11key/key.go

```diff
@@ -166,12 +204,18 @@ func (ps *Key) setup(privateKeyLabel string) (err error) {
 	}
 	ps.privateKeyHandle = privateKeyHandle
 
-	publicKey, err := getPublicKey(ps.module, session, privateKeyHandle)
+	publicKey, err := ps.getPublicKey(ps.module, session, privateKeyHandle)
 	if err != nil {
 		ps.module.CloseSession(session)
 		return
 	}
-	ps.publicKey = publicKey
+	if ps.keyType == EC {
+		t := publicKey.(ecdsa.PublicKey)
```

Comment:
I think you can delete this chunk of code if you change the return statements in `get{EC,RSA}PublicKey` to return a pointer.

I believe that allows you to drop the `keyType` member entirely. I'd recommend moving lines 245-264, in `getPrivateKey`, where you retrieve the key type, into their own function and calling that function from `getPublicKey`. Then `keyType` becomes a purely local variable, used only decide which of `getRSAPublicKey`, `getECPublicKey` to call.